File: reference/whenwords/src/parse-duration.test.ts
import { describe, test, expect } from "bun:test";
import { parseDuration } from "./parse-duration";

describe("parse-duration", () => {
  describe("compact format", () => {
    test.each([
      { input: "2h30m", expected: 9000 },
      { input: "2h 30m", expected: 9000 },
      { input: "2h, 30m", expected: 9000 },
      { input: "1.5h", expected: 5400 },
      { input: "90m", expected: 5400 },
      { input: "90min", expected: 5400 },
      { input: "45s", expected: 45 },
      { input: "45sec", expected: 45 },
      { input: "2d", expected: 172800 },
      { input: "1w", expected: 604800 },
      { input: "1d 2h 30m", expected: 95400 },
      { input: "2hr", expected: 7200 },
      { input: "2hrs", expected: 7200 },
      { input: "30mins", expected: 1800 },
    ])("$input → $expected", ({ input, expected }) => {
      expect(parseDuration(input)).toBe(expected);
    });
  });

  describe("verbose format", () => {
    test.each([
      { input: "2 hours 30 minutes", expected: 9000 },
      { input: "2 hours and 30 minutes", expected: 9000 },
      { input: "2 hours, and 30 minutes", expected: 9000 },
      { input: "2.5 hours", expected: 9000 },
      { input: "90 minutes", expected: 5400 },
      { input: "2 days", expected: 172800 },
      { input: "1 week", expected: 604800 },
      { input: "1 day, 2 hours, and 30 minutes", expected: 95400 },
      { input: "45 seconds", expected: 45 },
    ])("$input → $expected", ({ input, expected }) => {
      expect(parseDuration(input)).toBe(expected);
    });
  });

  describe("colon notation", () => {
    test.each([
      { input: "2:30", expected: 9000 },
      { input: "1:30:00", expected: 5400 },
      { input: "0:05:30", expected: 330 },
    ])("$input → $expected", ({ input, expected }) => {
      expect(parseDuration(input)).toBe(expected);
    });
  });

  describe("case insensitive and whitespace tolerant", () => {
    test("2H 30M → 9000", () => {
      expect(parseDuration("2H 30M")).toBe(9000);
    });

    test("extra whitespace", () => {
      expect(parseDuration("  2 hours   30 minutes  ")).toBe(9000);
    });
  });

  describe("error cases", () => {
    test("empty string", () => {
      expect(() => parseDuration("")).toThrow();
    });

    test("no units", () => {
      expect(() => parseDuration("hello world")).toThrow();
    });

    test("negative", () => {
      expect(() => parseDuration("-5 hours")).toThrow();
    });

    test("just a number", () => {
      expect(() => parseDuration("42")).toThrow();
    });

    test("unrecognized unit", () => {
      expect(() => parseDuration("5 foos")).toThrow();
    });
  });
});

File: research/experiments/sweep/runs/whenwords-full-cpp-sonnet45/tests/parse_duration_test.cpp
#include "doctest.h"
#include "parse_duration.h"
#include <stdexcept>

TEST_CASE("parseDuration - compact format") {
    CHECK(parseDuration("2h30m") == 9000);
    CHECK(parseDuration("2h 30m") == 9000);
    CHECK(parseDuration("2h, 30m") == 9000);
    CHECK(parseDuration("1.5h") == 5400);
    CHECK(parseDuration("90m") == 5400);
    CHECK(parseDuration("90min") == 5400);
    CHECK(parseDuration("45s") == 45);
    CHECK(parseDuration("45sec") == 45);
    CHECK(parseDuration("2d") == 172800);
    CHECK(parseDuration("1w") == 604800);
    CHECK(parseDuration("1d 2h 30m") == 95400);
    CHECK(parseDuration("2hr") == 7200);
    CHECK(parseDuration("2hrs") == 7200);
    CHECK(parseDuration("30mins") == 1800);
}

TEST_CASE("parseDuration - verbose format") {
    CHECK(parseDuration("2 hours 30 minutes") == 9000);
    CHECK(parseDuration("2 hours and 30 minutes") == 9000);
    CHECK(parseDuration("2 hours, and 30 minutes") == 9000);
    CHECK(parseDuration("2.5 hours") == 9000);
    CHECK(parseDuration("90 minutes") == 5400);
    CHECK(parseDuration("2 days") == 172800);
    CHECK(parseDuration("1 week") == 604800);
    CHECK(parseDuration("1 day, 2 hours, and 30 minutes") == 95400);
    CHECK(parseDuration("45 seconds") == 45);
}

TEST_CASE("parseDuration - colon notation") {
    CHECK(parseDuration("2:30") == 9000);
    CHECK(parseDuration("1:30:00") == 5400);
    CHECK(parseDuration("0:05:30") == 330);
}

TEST_CASE("parseDuration - case and whitespace tolerance") {
    CHECK(parseDuration("2H 30M") == 9000);
    CHECK(parseDuration("  2 hours   30 minutes  ") == 9000);
}

TEST_CASE("parseDuration - error cases") {
    CHECK_THROWS_AS(parseDuration(""), std::invalid_argument);
    CHECK_THROWS_AS(parseDuration("hello world"), std::invalid_argument);
    CHECK_THROWS_AS(parseDuration("-5 hours"), std::invalid_argument);
    CHECK_THROWS_AS(parseDuration("42"), std::invalid_argument);
    CHECK_THROWS_AS(parseDuration("5 foos"), std::invalid_argument);
}
